packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js

```diff
@@ -1,621 +1,944 @@
+const unreachableDeadline = MAX_UINT_VAL;
+
+// Common contract objects.
+let collateralToken;
+let syntheticToken;
+let emp;
+let mockOracle;
+let identifierWhitelist;
+let identifier;
+let finder;
+let store;
+let timer;
+let collateralWhitelist;
+let constructorParams;
+let currentTestIterationVersion;
+
+// Js Objects, clients and helpers
+let client;
+let convertCollateral;
+let convertSynthetic;
+let convertPrice;
+let dummyLogger;
+
+// Perpetual
+let configStore;
+let optimisticOracle;
+let fundingRateIdentifier;
+
+// Helper functions
+// TODO figure out the best pattern to refactor these into a library to make them re-usable in other tests.
+const updateAndVerify = async (client, expectedSponsors, expectedPositions) => {
+  await client.update();
+  assert.deepStrictEqual(client.getAllSponsors().sort(), expectedSponsors.sort());
+  assert.deepStrictEqual(client.getAllPositions().sort(), expectedPositions.sort());
+};
+
+// Set the funding rate and advances time by 10k seconds.
+const _setFundingRateAndAdvanceTime = async fundingRate => {
+  const currentTime = (await emp.getCurrentTime()).toNumber();
+  await emp.proposeFundingRate({ rawValue: fundingRate }, currentTime);
+  await emp.setCurrentTime(currentTime + 10000);
+};
+
+// If the current version being executed is part of the `supportedVersions` array then return `it` to run the test.
+// Else, do nothing. Can be used exactly in place of a normal `it` to parameterize contract types and versions supported
+// for a given test.eg: versionedIt(["Perpetual-latest"])("test name", async function () { assert.isTrue(true) })
+// Note that a second param can be provided to make the test an `it.only` thereby ONLY running that single test, on
+// the provided version. This is very useful for debugging and writing single unit tests without having ro run all tests.
+const versionedIt = function(supportedVersions, shouldBeItOnly = false) {
+  if (shouldBeItOnly) return runTestForContractVersion(supportedVersions) ? it.only : () => {};
+  return runTestForContractVersion(supportedVersions) ? it : () => {};
+};
+
+const runTestForContractVersion = function(supportedVersions) {
+  // Validate that the array of supportedVersions provided is in the SUPPORTED_CONTRACT_VERSIONS OR is any.
+  if ([...SUPPORTED_CONTRACT_VERSIONS, "any"].filter(value => supportedVersions.includes(value)).length == 0) {
+    throw new Error(
+      `Contract versioned specified ${supportedVersions} is not part of the supported contracts for this test suit`
+    );
+  }
+  // Return if the `currentTestIterationVersion` is part of the supported versions includes any. Returning true
+  // means that test will be run. Else, if returned false, the test will be skipped.
+  return supportedVersions.includes(currentTestIterationVersion) || supportedVersions.includes("any");
+};
+
+const _createConstructorParamsForContractVersion = async function(contractVersion, contractType) {
```

Comment:
Different versions require different constructors. this is useful in switching based on the version in question. Will also be pulled to common util.

Reply:
for sure. I will do that kind of validation within the util when I refactor this out.